### DISCUSSION TITLE
API

### DISCUSSION THREAD

**Discussion by pseudoparenchymatous** [nvim-neorocks/lz.n#18] [Q&A] [answered]
It says a UI is non-goal for this plugin. Maybe a separate extension plugin for it. But for now, to help debug a config, perhaps an API that can give a list of all plugins managed by `lz.n` and full spec of each plugin plus an `is_loaded` boolean?

**Comment by mrcjkb** [comment]
That sounds reasonable. I don't mind making lz.n extensible.

**Comment by BirdeeHub** [comment, marked as answer]
I made a handler that can do this for you

Kinda quick and dirty but this works.

Say we saved it at lua/lz_n_debug

```lua
---@type table<string, string>
local states = {}

local M = {
  ---@type lz.n.Handler
  handler = {
	-- normally we would use this to allow the user to pass us a spec item
	-- but here its just a name that we are going to set the value of ourselves
	spec_field = "is_loaded",
	-- del is called whenever require('lz.n').trigger_load is called, once for every plugin spec
	---@param plugin lz.n.Plugin
	del = function (plugin)
	  states[plugin.name] = plugin.name
	end,
	-- add is called whenever require('lz.n').load parses a spec
        -- we could use it to store the plugins as they are added, and simply update is_loaded with the del function too.
	add = function(_) end,
  },
}

function M.get_all_plugins()
  local result = vim.deepcopy(require("lz.n.state").plugins)
  for _, name in pairs(states) do
	if result[name] ~= nil then
	  ---@diagnostic disable-next-line: inject-field
	  result[name].is_loaded = true
	end
  end
  return result
end

function M.get_a_plugin(name)
  local result = vim.deepcopy(require("lz.n.state").plugins[name])
  if result ~= nil and states[name] ~= nil then
    ---@diagnostic disable-next-line: inject-field
    result.is_loaded = true
  end
  return result
end

return M
```

Simply add `require('lz.n').register_handler(require('lz_n_debug').handler)` to the start of your config

and then you can call `:lua print(vim.inspect(require('lz_n_debug').get_all_plugins()))` and it will print out all the specs and whether they were loaded or not or you can `:lua print(vim.inspect(require('lz_n_debug').get_a_plugin("apluginname")))` for just 1

This works because the handler's del function is called with the plugin spec every time `require('lz.n').trigger_load` is called

You could get them all via add and mark them loaded when del is called, and is probably the better route if you were to make a whole UI for it, but i took a shortcut and just queried lz.n.state so that i just have to store the names

**Reply by pseudoparenchymatous** [reply]
Thanks!

**Reply by BirdeeHub** [reply]
No problem :) I made lz.n hackable with mrcjkb's guidance so its probably also on me to tell people how to use that while the plugin is still new XD Pretty happy that it ended up as such a simple interface while still having all the capabilies of being a built in handler type.

**Reply by BirdeeHub** [reply]
For anyone wondering, this is the other way I referred to

```lua
---@class lz-n.Plugin: lz.n.Plugin
---@field is_loaded? boolean

---@type table<string, lz-n.Plugin>
local states = {}

local M = {
  ---@type lz.n.Handler
  handler = {
    -- this field does nothing but it does stop others from using is_loaded,
    -- which is good because we are going to write to it.
    spec_field = "is_loaded",
    ---@param plugin lz-n.Plugin
    del = function (plugin)
      if not states[plugin.name] then
        states[plugin.name] = plugin
      end
      states[plugin.name].is_loaded = true
    end,
    ---@param plugin lz-n.Plugin
    add = function(plugin)
      states[plugin.name] = plugin
      if plugin.lazy then
        states[plugin.name].is_loaded = false
      else
        states[plugin.name].is_loaded = true
      end
    end,
  },
}

function M.get_all_plugins()
  return states
end

function M.get_a_plugin(name)
  return states[name]
end

return M
```
